python/tank/platform/engine.py

```diff
@@ -565,63 +566,52 @@ def execute_in_main_thread(self, func, *args, **kwargs):
         Note, this currently only works if Qt is available, otherwise it just
         executes on the current thread.
 
-        # ---------------------------------------------------------------------------------
-        # The following test checks that execute_in_main_thread is itself thread-safe!  It
-        # runs a simple test a number of times in multiple threads and ensures the result
-        # returned is as expected.  If this isn't the case it will print out one or more:
-        #     'Result mismatch...'
-        #
-        import threading
-        import random
-        import time
-        import sgtk
-        from sgtk.platform.qt import QtCore, QtGui
-
-        NUM_TEST_THREADS=20
-        NUM_THREAD_ITERATIONS=30
-
-        def run_in_main_thread(v):
-            print "Running", v
-            if QtCore.QThread.currentThread() != QtGui.QApplication.instance().thread():
-                print " > but not running in main thread!"
-            return v
-
-        def threaded_work(val):
-            eng = sgtk.platform.current_engine()
-            c_time = 0.0
-            for i in range(NUM_THREAD_ITERATIONS):
-                time.sleep(random.randint(0, 10)/10.0)
-                arg = (val, i)
-                st = time.time()
-                ret_val = eng.execute_in_main_thread(run_in_main_thread, arg)
-                e = time.time()
-                #print "Time to run: %0.4fs" % (e-st)
-                c_time += (e-st)
-                if ret_val != arg:
-                    print "Result mismatch: %s != %s!!" % (ret_val, arg)
-
-            print "Cumulative time for thread %d: %0.4fs" % (val, c_time)
-
-        threads = []
-        for ti in range(NUM_TEST_THREADS):
-            t = threading.Thread(target=lambda:threaded_work(ti))
-            t.start()
-            threads.append(t)
-        # ---------------------------------------------------------------------------------
-        # ---------------------------------------------------------------------------------
-
         :param func: function to call
         :param args: arguments to pass to the function
         :param kwargs: named arguments to pass to the function
 
         :returns: the result of the function call
         """
-        if self._invoker:
+        return self._execute_in_main_thread("_invoker", func, *args, **kwargs)
```

Comment:
I think we should add some warnings to this method to make it very clear that it may give rise to deadlocks and other issues, and describe when this type of situation can occur. I am guessing there are situations when you really want to do something and get a return value, so i am guessing the method is still needed, but we should make it clear what the risks are.